src/sshagent/OpenSSHKey.cpp

```diff
@@ -0,0 +1,481 @@
+        m_error = tr("Key file way too small.");
+        return false;
+    }
+
+    if (QString::fromLatin1(magic) != "openssh-key-v1") {
+        m_error = tr("Key file magic header id invalid");
+        return false;
+    }
+
+    stream.readString(m_cipherName);
+    stream.readString(m_kdfName);
+    stream.readString(m_kdfOptions);
+
+    quint32 numberOfKeys;
+    stream.read(numberOfKeys);
+
+    if (numberOfKeys == 0) {
+        m_error = tr("Found zero keys");
+        return false;
+    }
+
+    for (quint32 i = 0; i < numberOfKeys; i++) {
+        QByteArray publicKey;
+        if (!stream.readString(publicKey)) {
+            m_error = tr("Failed to read public key.");
+            return false;
+        }
+
+        if (i == 0) {
+            BinaryStream publicStream(&publicKey);
+            if (!readPublic(publicStream)) {
+                return false;
+            }
+        }
+    }
+
+    // padded list of keys
+    if (!stream.readString(m_rawPrivateData)) {
+        m_error = tr("Corrupted key file, reading private key failed");
+        return false;
+    }
+
+    // load private if no encryption
+    if (!encrypted()) {
+        return decrypt();
+    }
+
+    return true;
+}
+
+bool OpenSSHKey::encrypted() const
+{
+    return (m_cipherName != "none");
+}
+
+bool OpenSSHKey::decrypt(const QString& passphrase)
+{
+    QScopedPointer<SymmetricCipher> cipher;
+
+    if (!m_privateData.isEmpty()) {
+        return true;
+    }
+
+    if (m_rawPrivateData.isEmpty()) {
+        m_error = tr("No private key payload to decrypt");
+        return false;
+    }
+
+    if (m_cipherName == "aes256-cbc") {
+        cipher.reset(new SymmetricCipher(SymmetricCipher::Aes256, SymmetricCipher::Cbc, SymmetricCipher::Decrypt));
+    } else if (m_cipherName == "aes256-ctr") {
+        cipher.reset(new SymmetricCipher(SymmetricCipher::Aes256, SymmetricCipher::Ctr, SymmetricCipher::Decrypt));
+    } else if (m_cipherName != "none") {
+        m_error = tr("Unknown cipher: ") + m_cipherName;
+        return false;
+    }
+
+    if (m_kdfName == "bcrypt") {
+        if (!cipher) {
+            m_error = tr("Trying to run KDF without cipher");
+            return false;
+        }
+
+        BinaryStream optionStream(&m_kdfOptions);
+
+        QByteArray salt;
+        quint32 rounds;
+
+        optionStream.readString(salt);
+        optionStream.read(rounds);
+
+        QByteArray decryptKey;
+        decryptKey.fill(0, cipher->keySize() + cipher->blockSize());
+
+        QByteArray phraseData = passphrase.toLatin1();
+        if (bcrypt_pbkdf(phraseData, salt, decryptKey, rounds) < 0) {
+            m_error = tr("Decryption failed, wrong passphrase?");
+            return false;
+        }
+
+        QByteArray keyData, ivData;
+        keyData.setRawData(decryptKey.data(), cipher->keySize());
+        ivData.setRawData(decryptKey.data() + cipher->keySize(), cipher->blockSize());
+
+        cipher->init(keyData, ivData);
```

Comment:
I still find it peculiar, that the OpenSSH format has no HMAC for the cipher text. PuTTY format has that.

Reply:
You want me to file an issue with OpenSSH developers? 😕

Reply:
No, that was just a comment on the sidelines.